src/modifiercontext.ts

```diff
@@ -21,12 +21,74 @@ import { ChordSymbol } from './chordsymbol';
 import { Bend } from './bend';
 import { Vibrato } from './vibrato';
 
+export interface ModifierContextState {
+  right_shift: number;
+  left_shift: number;
+  text_line: number;
+  top_text_line: number;
+}
+
+export interface ModifierContextMetrics {
+  width: number;
+  spacing: number;
+}
+
+export type PreformatModifierType =
+  | typeof StaveNote
+  | typeof Dot
+  | typeof FretHandFinger
+  | typeof Accidental
+  | typeof Stroke
+  | typeof GraceNoteGroup
+  | typeof NoteSubGroup
+  | typeof StringNumber
+  | typeof Articulation
+  | typeof Ornament
+  | typeof Annotation
+  | typeof ChordSymbol
+  | typeof Bend
+  | typeof Vibrato;
+
+export type PostformatModifierType = typeof StaveNote;
+
+export type ModifierClass =
```

Comment:
There must be a better approach here -- maybe there's a required method that all of them implement? (format / preformat?)

Reply:
I think better to create an interface in common that has `getCategory` and `setModifierContext`

Reply:
As I said above:
>Reason the mix between StaveNote and Modifiers in ModifierClass this is worth a refactoring.

Reply:
Sorry -- I don't understand what you mean by "mix between StaveNote and Modifiers". Can't we have:

```typescript
export type ModifierClass = Modifer | StaveNote
interface Preformattable {
  format(...)
}

interface Postformattable {
  postformat(...)
}
```

Reply:
Of course we can compile almost the mix we want. It is more a conceptual and maintinability topic. Perhaps I am missing something, but I understand that `StaveNote` is not a `Modifier`. 
1) My concerns related to `StaveNote`:
- If I am right `StaveNote` should not be a `ModifierClass`
- Using `addModifier` and `getModifiers` with `StaveNote` does not make a lot of sense.
2) The mix you propose will not work due to the fact the modifier does not provide the `format` so you have to make the complete list: `accidental, annotation` surprisingly (for me) `Tremolo` is a `Modifier` but it is not in the list and does not provide the static method `format`
3) `Modifier` does not provide `postFormat`

We can leave it as it is as well, this is your choice. :)

Reply:
Thanks, rereading this, and going through the rest of the code:

* `ModifierContext` is not just modifiers, it's a container for all elements that belong within the inner formatting contexts within the tick contexts, so it includes tickables (like `StaveNote`, `TabNote`, etc.)
* the `format()` calls here refer to the static class methods for the specific elements, which all modifiers in `PREFORMAT` should have. Since interfaces apply to class instances, not the classes themselves, we can't really create a generic type for them.
* `postFormat` is also static, so we can't really create a generic type for it too.

So I'd say leave this class as an exception and use `any[]` for `PREFORMAT`, `POSTFORMAT`.

Re: `ModifierClass`, I don't understand why it's there, why can't we just use `Modifier` here? Do we have any code that adds `StaveNote` as a modifier?

(FYI: https://github.com/0xfe/vexflow/wiki/How-Formatting-Works)

Reply:
I was wrong `ModifierClass = Modifier | StaveNote` works and does not produce a failure because `Tremolo` is not in `PreformatModifierType`. This together with `as` in `StaveNote.getAccidentals` and `StaveNote.getDots`makes the change more sensible.
> Re: `ModifierClass`, I don't understand why it's there, why can't we just use `Modifier` here? Do we have any code that adds `StaveNote` as a modifier?

`StaveNote` does it:
https://github.com/0xfe/vexflow/blob/ba717d830c8a37d25f4e690f9b2983adc852d400/src/stavenote.ts#L877
 
> (FYI: https://github.com/0xfe/vexflow/wiki/How-Formatting-Works)

According to the wikie: "`Modifiers` do not, but instead have a static `<Modifier>.format()` method."
However `Tremolo` is a `Modifier` and does not have the static method `format`. **Why is that?**

Reply:
Okay, so that code does not add `StaveNote` as a modifier, just as an element in the `ModifierContext` -- this is the same for `TabNote` and anything that has `setModifierContext`.

So two things, for `ModifierClass`, let's rename it to something more descriptive, maybe `ModifierContextMember` as so:

```typescript
export interface ModifierContextMember {
  getCategory(...) : string;
  setModifierContext(...);
}
```

Then lets fix some of the internal comments and variable names: `modifier -> member` so it's clear that these are not just modifiers.

> However `Tremolo` is a `Modifier` and does not have the static method `format`.

This is probably because no additional reformatting is required for `Tremolo` --  notice that it's also not in `PREFORMAT`.

Reply:
This is a tricky one -- that's for digging in.

Reply:
Done